File: 백준/Platinum/5670. 휴대폰 자판/휴대폰 자판.py
import sys
input = lambda : sys.stdin.readline().strip()

def make():
    trie = {}
    words = [input() for _ in range(int(input()))]
    
    for word in words: 
        tmp = trie
        for i in word:
            if i not in tmp:
                tmp[i] = {}
            tmp = tmp[i]
        tmp[0] = 1

    result = 0
    for word in words:
        tmp = trie[word[0]]
        result+=1
        for i in word[1:]:
            if len(tmp.keys())!=1:
                result+=1
            tmp = tmp[i]
            # print(result)
    a = round(result/len(words),2)
    print(f'{a:0.2f}')

while True:
    try: 
        make()
    except:
        break

File: 백준/Platinum/5670. 휴대폰 자판/휴대폰 자판.java
import java.io.*;
import java.util.*;

class TrieNode {
    Map<Character, TrieNode> children = new HashMap<>();
    boolean isEndOfWord;
}

class Trie {
    private final TrieNode root;
    
    public Trie() {
        root = new TrieNode();
    }
    
    public void insert(String word) {
        TrieNode node = root;
        for (char ch : word.toCharArray()) {
            node.children.putIfAbsent(ch, new TrieNode());
            node = node.children.get(ch);
        }
        node.isEndOfWord = true;
    }
    
    public int countPresses(String word) {
        TrieNode node = root;
        int presses = 1; // 첫 글자는 무조건 입력해야 함
        node = node.children.get(word.charAt(0));
        
        for (int i = 1; i < word.length(); i++) {
            if (node.children.size() != 1 || node.isEndOfWord) {
                presses++;
            }
            node = node.children.get(word.charAt(i));
        }
        return presses;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        
        String line;
        while ((line = br.readLine()) != null && !line.isEmpty()) {
            int n = Integer.parseInt(line);
            Trie trie = new Trie();
            List<String> words = new ArrayList<>();
            
            for (int i = 0; i < n; i++) {
                String word = br.readLine();
                words.add(word);
                trie.insert(word);
            }
            
            int totalPresses = 0;
            for (String word : words) {
                totalPresses += trie.countPresses(word);
            }
            
            double average = (double) totalPresses / n;
            sb.append(String.format("%.2f\n", average));
        }
        System.out.print(sb);
    }
}